consensus/progpow/algorithm_progpow.go

```diff
@@ -0,0 +1,431 @@
+	var randState kiss99State
+
+	fnvHash := uint32(0x811c9dc5)
+
+	randState.z = fnv1a(&fnvHash, lower32(seed))
+	randState.w = fnv1a(&fnvHash, higher32(seed))
+	randState.jsr = fnv1a(&fnvHash, lower32(seed))
+	randState.jcong = fnv1a(&fnvHash, higher32(seed))
+
+	// Create a random sequence of mix destinations for merge()
+	// and mix sources for cache reads
+	// guarantees every destination merged once
+	// guarantees no duplicate cache reads, which could be optimized away
+	// Uses Fisher-Yates shuffle
+	var dstSeq = [32]uint32{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31}
+	var srcSeq = [32]uint32{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31}
+
+	for i := uint32(progpowRegs - 1); i > 0; i-- {
+		j := kiss99(&randState) % (i + 1)
+		dstSeq[i], dstSeq[j] = dstSeq[j], dstSeq[i]
+		j = kiss99(&randState) % (i + 1)
+		srcSeq[i], srcSeq[j] = srcSeq[j], srcSeq[i]
+	}
+	return randState, dstSeq, srcSeq
+}
+
+// Random math between two input values
+func progpowMath(a uint32, b uint32, r uint32) uint32 {
+	switch r % 11 {
+	case 0:
+		return a + b
+	case 1:
+		return a * b
+	case 2:
+		return higher32(uint64(a) * uint64(b))
+	case 3:
+		if a < b {
+			return a
+		}
+		return b
+	case 4:
+		return rotl32(a, b)
+	case 5:
+		return rotr32(a, b)
+	case 6:
+		return a & b
+	case 7:
+		return a | b
+	case 8:
+		return a ^ b
+	case 9:
+		return uint32(bits.LeadingZeros32(a) + bits.LeadingZeros32(b))
+	case 10:
+		return uint32(bits.OnesCount32(a) + bits.OnesCount32(b))
+
+	default:
+		return 0
+	}
+}
+
+func progpowLoop(seed uint64, loop uint32, mix *[progpowLanes][progpowRegs]uint32,
+	lookup func(index uint32) []byte,
+	cDag []uint32, datasetSize uint32) {
+	// All lanes share a base address for the global load
+	// Global offset uses mix[0] to guarantee it depends on the load result
+	gOffset := mix[loop%progpowLanes][0] % (64 * datasetSize / (progpowLanes * progpowDagLoads))
+
+	var (
+		srcCounter uint32
+		dstCounter uint32
+		randState  kiss99State
+		srcSeq     [32]uint32
+		dstSeq     [32]uint32
+		rnd        = kiss99
+		index      uint32
+		data_g     []uint32 = make([]uint32, progpowDagLoads)
+	)
+	// 256 bytes of dag data
+	dag_item := make([]byte, 256)
+	// The lookup returns 64, so we'll fetch four items
+	copy(dag_item, lookup((gOffset*progpowLanes)*progpowDagLoads))
+	copy(dag_item[64:], lookup((gOffset*progpowLanes)*progpowDagLoads+16))
+	copy(dag_item[128:], lookup((gOffset*progpowLanes)*progpowDagLoads+32))
+	copy(dag_item[192:], lookup((gOffset*progpowLanes)*progpowDagLoads+48))
+
+	// Lanes can execute in parallel and will be convergent
+	for l := uint32(0); l < progpowLanes; l++ {
+
+		// initialize the seed and mix destination sequence
+		randState, dstSeq, srcSeq = progpowInit(seed)
+		srcCounter = uint32(0)
+		dstCounter = uint32(0)
+
+		//if progpowCntCache > progpowCntMath {
```

Comment:
commented code ? we need it?

Reply:
I didn't write this file or even see that line of code, lol. I'll remove it though, because I agree we don't need dead code :smile:

Reply:
removed